File: cses/08_string_algorithms/1731_word-combinations/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        sys.stdout.write("0\n")
        return

    s = data[0]                       # dạng bytes; s[j] trả về giá trị byte (int)
    k = int(data[1]) if len(data) > 1 else 0
    words = data[2:2 + k]
    n = len(s)
    MOD = 1000000007

    # Lưu Trie trong một dict chung: khóa = (node << 7) | char -> chỉ số nút con.
    # is_word[v] = 1 nếu nút v là điểm kết thúc của một từ trong từ điển.
    trie = {}
    is_word = bytearray()
    is_word.append(0)                 # nút gốc mang chỉ số 0
    node_count = 1
    for w in words:
        if len(w) > n:                # từ dài hơn s không bao giờ khớp -> bỏ qua
            continue
        v = 0
        for c in w:
            key = (v << 7) | c
            nxt = trie.get(key)
            if nxt is None:           # tạo nút con mới nếu chưa tồn tại
                nxt = node_count
                node_count += 1
                trie[key] = nxt
                is_word.append(0)
            v = nxt
        is_word[v] = 1                # đánh dấu kết thúc từ

    # dp[i] = số cách ghép được tiền tố s[0..i-1]; dp[0] = 1 (chuỗi rỗng).
    dp = [0] * (n + 1)
    dp[0] = 1

    trie_get = trie.get               # gán cục bộ để tăng tốc
    isw = is_word
    ss = s
    for i in range(n):
        di = dp[i]
        if not di:
            continue
        di %= MOD
        # Đi trên Trie theo các ký tự s[i..]; mỗi khi gặp nút kết thúc từ
        # (tức s[i..j] là một từ) thì cộng dp[i] vào dp[j+1].
        v = 0
        j = i
        while j < n:
            nxt = trie_get((v << 7) | ss[j])
            if nxt is None:           # không còn nhánh khớp -> dừng
                break
            v = nxt
            if isw[v]:
                dp[j + 1] += di       # di < MOD nên cộng dồn an toàn (< ~5e12)
            j += 1

    sys.stdout.write(str(dp[n] % MOD) + "\n")


main()

File: cses/08_string_algorithms/1731_word-combinations/solution.cpp
#include <array>
#include <iostream>
#include <string>
#include <vector>

namespace {

constexpr int MOD = 1'000'000'007;

// Mỗi nút Trie: next[c] là chỉ số nút con theo ký tự c (0 nghĩa là chưa có),
// is_word đánh dấu nút này là điểm kết thúc của một từ trong từ điển.
struct TrieNode {
    std::array<int, 26> next{};
    bool is_word = false;
};

}  // namespace

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string target;
    int word_count = 0;
    std::cin >> target >> word_count;

    const int target_length = static_cast<int>(target.size());
    std::vector<TrieNode> trie(1);  // nút gốc mang chỉ số 0
    trie.reserve(1'000'001);

    // Xây Trie từ toàn bộ từ điển; bỏ qua từ dài hơn s vì không thể khớp.
    for (int index = 0; index < word_count; ++index) {
        std::string word;
        std::cin >> word;
        if (word.size() > target.size()) {
            continue;
        }

        int node = 0;
        for (const char character : word) {
            const int letter = character - 'a';
            if (trie[node].next[letter] == 0) {  // tạo nút con nếu chưa tồn tại
                trie[node].next[letter] = static_cast<int>(trie.size());
                trie.emplace_back();
            }
            node = trie[node].next[letter];
        }
        trie[node].is_word = true;  // đánh dấu kết thúc từ
    }

    // ways[i] = số cách ghép được tiền tố s[0..i-1]; ways[0] = 1 (chuỗi rỗng).
    std::vector<int> ways(target_length + 1, 0);
    ways[0] = 1;

    for (int start = 0; start < target_length; ++start) {
        if (ways[start] == 0) {
            continue;
        }

        // Đi trên Trie theo các ký tự s[start..], mỗi khi gặp nút kết thúc từ
        // (tức s[start..end] là một từ) thì cộng dồn ways[start] vào ways[end+1].
        int node = 0;
        for (int end = start; end < target_length; ++end) {
            const int letter = target[end] - 'a';
            node = trie[node].next[letter];
            if (node == 0) {  // không còn nhánh khớp -> dừng
                break;
            }
            if (trie[node].is_word) {
                ways[end + 1] += ways[start];
                if (ways[end + 1] >= MOD) {
                    ways[end + 1] -= MOD;
                }
            }
        }
    }

    std::cout << ways[target_length] << '\n';
    return 0;
}
